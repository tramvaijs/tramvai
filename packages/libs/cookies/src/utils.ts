import type { CookieOptions } from './cookies.h';

const DEFAULT_PATH = '/';
const MS_IN_DAY = 864e5;

/**
 * Apply default value for unspecified options
 */
export function applyDefaults(
  options: CookieOptions = {},
  defaults: CookieOptions = {}
): CookieOptions {
  const expires = options.expires || defaults.expires;
  const domain = options.domain || defaults.domain;
  const path = options.path ?? defaults.path ?? DEFAULT_PATH;
  const secure = options.secure ?? defaults.secure;
  const httpOnly = options.httpOnly ?? defaults.httpOnly;
  const sameSite = options.sameSite ?? defaults.sameSite;

  return {
    expires,
    domain,
    path,
    secure,
    httpOnly,
    sameSite,
  };
}

/**
 * Determine cookie expiration date
 * If succesful the result will be a valid Date, otherwise it will be an invalid Date or false(ish)
 */
export function getExpireDate(expires?: number | string | Date): Date | null {
  const expireDate = expires
    ? new Date(
        // in case expires is an integer, it should specify the number of days till the cookie expires
        typeof expires === 'number'
          ? new Date().getTime() + expires * MS_IN_DAY
          : // else expires should be either a Date object or in a format recognized by Date.parse()
            expires
      )
    : null;

  return expireDate;
}
