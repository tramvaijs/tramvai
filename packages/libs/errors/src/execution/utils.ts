import { ExecutionAbortError } from './executionAbortError';
import { ExecutionError } from './executionError';
import { ExecutionTimeoutError } from './executionTimeoutError';

export const isExecutionError = (err: Error): err is ExecutionError => {
  return err.name === ExecutionError.errorName;
};

export const isExecutionAbortError = (err: Error): err is ExecutionAbortError => {
  return err.name === ExecutionAbortError.errorName || err.name === ExecutionTimeoutError.errorName;
};

export const isExecutionTimeoutError = (err: Error): err is ExecutionTimeoutError => {
  return err.name === ExecutionTimeoutError.errorName;
};
