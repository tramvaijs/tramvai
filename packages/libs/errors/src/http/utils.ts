import { HttpError } from './httpError';
import { NotFoundError } from './notFound';
import { RedirectFoundError } from './redirect';

export const throwNotFoundError = (parameters?: ConstructorParameters<typeof NotFoundError>[0]) => {
  throw new NotFoundError(parameters);
};

export const isNotFoundError = (err: Error): err is NotFoundError => {
  return err.name === NotFoundError.errorName;
};

export const throwRedirectFoundError = (
  parameters: ConstructorParameters<typeof RedirectFoundError>[0]
) => {
  throw new RedirectFoundError(parameters);
};

export const isRedirectFoundError = (err: Error): err is RedirectFoundError => {
  return err.name === RedirectFoundError.errorName;
};

export const throwHttpError = (parameters: ConstructorParameters<typeof HttpError>[0]) => {
  throw new HttpError(parameters);
};

export const isHttpError = (err: Error): err is HttpError => {
  return err.name === HttpError.errorName || isNotFoundError(err) || isRedirectFoundError(err);
};
