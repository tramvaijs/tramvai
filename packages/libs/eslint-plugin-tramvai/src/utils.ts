import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { execSync } from 'node:child_process';

import findCacheDir from 'find-cache-dir';
import browserslist from 'browserslist';
import { ICruiseResult } from 'dependency-cruiser';

import { disabledRules, featureToRule } from './rulesMap';
import { Rules, TramvaiConfig } from './types';

export function getLibraryEslintConfig(libraryName: string, projectRoot?: string) {
  const libraryRoot = path.dirname(
    require.resolve(libraryName, projectRoot ? { paths: [projectRoot] } : {})
  );

  try {
    return require(`${libraryRoot}/eslint.js`);
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error('Failed to parse custom settings config!');

    return [];
  }
}

export function resolveBrowserslistConfig(projectPath: string): any {
  return browserslist.loadConfig({ path: projectPath });
}

export function getUsedPolyfillsPathsFromLibrary(libraryPath: string): string[] {
  const libraryPolyfillImportsFile = path.join(libraryPath, 'imports.json');

  try {
    return JSON.parse(fs.readFileSync(libraryPolyfillImportsFile, 'utf-8'));
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error(`Failed to get polyfill library `);
    return [];
  }
}

const defaultPolyfillEntryPath = './src/polyfill.ts';
const defaultModernPolyfillEntryPath = './src/modern.polyfill.ts';

export function getUsedPolyfillsPathsFromProject(projectPath: string) {
  const tramvaiConfig = getTramvaiConfig(projectPath);
  const { projects } = tramvaiConfig;

  const applicationConfig = Object.values(projects).find(
    (projectConfig: { type: string }) => projectConfig.type === 'application'
  );
  const { polyfill, modernPolyfill } = applicationConfig;

  const polyfillEntryPath = path.join(projectPath, polyfill ?? defaultPolyfillEntryPath);
  const modernPolyfillEntryPath = path.join(
    projectPath,
    modernPolyfill ?? defaultModernPolyfillEntryPath
  );

  const polyfillImports = [];

  if (fs.existsSync(polyfillEntryPath)) {
    polyfillImports.push(...getImports(polyfillEntryPath, projectPath));
  } else {
    const defaultPolyfillsPath = require.resolve('@tinkoff/pack-polyfills');
    polyfillImports.push(...getImports(defaultPolyfillsPath, projectPath));
  }

  if (fs.existsSync(modernPolyfillEntryPath)) {
    polyfillImports.push(...getImports(modernPolyfillEntryPath, projectPath));
  }

  return polyfillImports;
}

export function patchRules(customPolyfillsPaths: string[], rules: Rules) {
  customPolyfillsPaths.forEach((polyfillPath) => {
    let polyfillName = polyfillPath;

    if (polyfillPath.includes('/core-js/modules/')) {
      polyfillName = path.parse(polyfillPath).name;
    }

    const polyfillRule = featureToRule(polyfillName);

    if (polyfillRule) {
      if (Array.isArray(polyfillRule)) {
        polyfillRule.forEach((ruleName) => {
          rules[ruleName] = 0;
        });
      } else {
        rules[polyfillRule] = 0;
      }
    }
  });

  disabledRules.forEach((ruleName) => {
    rules[ruleName] = 0;
  });
}

function getImports(filePath: string, projectRootPath: string) {
  const packageJSONPath = path.join(projectRootPath, 'package.json');
  const packageJSONHash = getFileContentHash(packageJSONPath);

  const baseCachePath =
    findCacheDir({ name: 'tramvai', cwd: projectRootPath }) ?? 'node_modules/.tramvai';
  const fileContentHash = getFileContentHash(filePath);
  const cachePath = path.join(baseCachePath, `${packageJSONHash}_${fileContentHash}`);

  if (fileContentHash && fs.existsSync(cachePath)) {
    return JSON.parse(fs.readFileSync(cachePath, 'utf-8'));
  }
  const allModules = cruiseSync(filePath);
  const imports = allModules
    .map((item) => item.source)
    .filter((item) => !item.includes('/core-js/internals/'));
  fs.mkdirSync(path.dirname(cachePath), { recursive: true });
  fs.writeFileSync(cachePath, JSON.stringify(imports));

  return imports;
}

function getFileContentHash(filePath: string) {
  try {
    const fileBuffer = fs.readFileSync(filePath);
    const hash = crypto.createHash('md5');
    hash.update(fileBuffer);

    return hash.digest('hex').slice(0, 6);
  } catch (_err) {}
}

function safeLoadFile(filePath: string | undefined) {
  if (!filePath) {
    return '';
  }

  try {
    return fs.readFileSync(filePath, 'utf-8');
  } catch (err: any) {
    if (err.code !== 'ENOENT') {
      // eslint-disable-next-line no-console
      console.error(`Failed to read ${filePath}`);
    }

    return '';
  }
}

function cruiseSync(filePath: string): ICruiseResult['modules'] {
  const output = execSync(`npx depcruise ${filePath} --no-config --output-type json`, {
    encoding: 'utf-8',
    maxBuffer: 8192 * 8192,
  });

  return JSON.parse(output).modules;
}

function getTramvaiConfig(projectPath: string) {
  try {
    const tramvaiConfig = fs.readFileSync(`${projectPath}/tramvai.json`, 'utf-8');
    return JSON.parse(tramvaiConfig) as TramvaiConfig;
  } catch (err: any) {
    throw new Error(`Tramvai config cannot be loaded: ${err}`);
  }
}
