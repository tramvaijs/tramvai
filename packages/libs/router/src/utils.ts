import T from '@tinkoff/utils/function/T';
import type { Url } from '@tinkoff/url';
import type { NavigateOptions, NavigationHook, NavigationSyncHook, NavigationGuard } from './types';

export const isFilePath = (pathname: string) => {
  return /\/.+\.[^/]+$/.test(pathname);
};

export const normalizeTrailingSlash = (pathname: string, trailingSlash = false) => {
  const hasTrailingSlash = pathname.endsWith('/');

  if (trailingSlash) {
    return hasTrailingSlash || isFilePath(pathname) ? pathname : `${pathname}/`;
  }

  return pathname.length > 1 && hasTrailingSlash ? pathname.slice(0, -1) : pathname;
};

export const normalizeManySlashes = (hrefOrPath: string) => {
  const [href, ...search] = hrefOrPath.split('?');

  return [href.replace(/\/+/g, '/').replace(/^(\w+):\//, '$1://'), ...search].join('?');
};

export const isSameHost =
  typeof window === 'undefined'
    ? T
    : (url: Url | URL) => {
        return !url.host || url.host === window.location.host;
      };

export const makeNavigateOptions = (options: string | NavigateOptions): NavigateOptions => {
  if (typeof options === 'string') {
    return { url: options };
  }

  return options;
};

export const registerHook = <T extends NavigationHook | NavigationSyncHook | NavigationGuard>(
  hooksSet: Set<T>,
  hook: T
) => {
  hooksSet.add(hook);

  return () => {
    hooksSet.delete(hook);
  };
};
