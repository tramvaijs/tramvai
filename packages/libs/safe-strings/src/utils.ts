export * from './safeStringify';
export * from './safeParseJSON';
export * from './safeStringifyJSON';
export * from './removeXss';
export * from './encode';
export * from './encodeForHTMLContext';
export * from './encodeForJSContext';
export * from './error';
