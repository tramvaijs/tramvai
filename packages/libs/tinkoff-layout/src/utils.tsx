import flatten from '@tinkoff/utils/array/flatten';
import type { ReactComponent, Wrapper, LayoutOptions } from './types.h';

const toArray = <T,>(arr: T | T[]) => (Array.isArray(arr) ? arr : [arr]);

export const composeComponent = (
  Component: ReactComponent,
  wrappers: Wrapper | Wrapper[]
): ReactComponent => {
  return toArray(wrappers ?? []).reduce((Wrapped, wrapper) => wrapper(Wrapped), Component);
};

export const composeLayoutOptions = (list: LayoutOptions[]): Required<LayoutOptions> => {
  return flatten<Required<LayoutOptions>>(list ?? []).reduce(
    (acc, item) => {
      const { components, wrappers } = acc;
      if (item.components) {
        Object.assign(components, item.components);
      }
      if (item.wrappers) {
        Object.keys(item.wrappers).forEach((component) => {
          if (!wrappers[component]) {
            wrappers[component] = [];
          }
          (wrappers[component] as Wrapper[]).push(...toArray(item.wrappers[component]));
        });
      }
      return acc;
    },
    {
      components: {},
      wrappers: {},
    }
  );
};
