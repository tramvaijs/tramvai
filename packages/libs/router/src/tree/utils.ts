import map from '@tinkoff/utils/array/map';
import {
  PARAMETER_DELIMITER,
  HISTORY_FALLBACK_REGEXP,
  WILDCARD_REGEXP,
  PARAM_PARSER_REGEXP,
  PARAM_MATCH_REGEXP,
} from './constants';
import type { Parameter, ParsedPath } from './types';
import { PartType } from './types';
import type { Params } from '../types';
import { normalizeTrailingSlash } from '../utils';

export const getParts = (pathname: string) =>
  pathname
    .split('/')
    .slice(pathname.startsWith('/') ? 1 : 0, pathname.endsWith('/') ? -1 : Infinity);

export const isHistoryFallback = (part: string) => {
  return HISTORY_FALLBACK_REGEXP.test(part);
};

export const isWildcard = (part: string) => {
  return WILDCARD_REGEXP.test(part);
};

export const isParameterized = (part: string) => {
  return !!part.match(PARAM_MATCH_REGEXP);
};

const parseParameter = (part: string): Parameter => {
  const [prefix = '', param, postfix = ''] = part.split(PARAMETER_DELIMITER);

  const match = PARAM_PARSER_REGEXP.exec(param);

  if (!match) {
    throw new Error('parameters should satisfy pattern "prefix:paramName(regexp)\\?:postfix"');
  }

  const [, paramName, regexp, optional] = match;
  const useRegexp = prefix || postfix || regexp;

  return {
    type: PartType.parameter,
    paramName,
    regexp: useRegexp
      ? new RegExp(`^${prefix}(${regexp || '.+'})${optional ? '?' : ''}${postfix}$`)
      : undefined,
    optional: !!optional && !prefix && !postfix,
  };
};

export const parse = (part: string): ParsedPath => {
  if (isHistoryFallback(part)) {
    return { type: PartType.historyFallback };
  }

  if (isWildcard(part)) {
    return { type: PartType.wildcard };
  }

  if (isParameterized(part)) {
    return parseParameter(part);
  }

  return { type: PartType.literal, value: part };
};

export const makePath = (pathname: string, params: Params) => {
  const parts = getParts(pathname);

  const result = map((part: string) => {
    if (isHistoryFallback(part) || isWildcard(part)) {
      throw new Error(
        `Pathname should be only a string with dynamic parameters, not a special string, got ${pathname}`
      );
    }

    if (isParameterized(part)) {
      const [prefix = '', param = '', postfix = ''] = part.split(PARAMETER_DELIMITER);
      const match = PARAM_PARSER_REGEXP.exec(param);

      if (!match) {
        throw new Error('parameters should satisfy pattern "prefix:paramName(regexp)\\?:postfix"');
      }

      const [, paramName, regexp, optional] = match;
      const value = params[paramName]?.toString();

      if (optional && !value) {
        return '';
      }

      if (!value) {
        throw new Error(`value for parameter for ${paramName} should be defined in params`);
      }

      if (regexp && !new RegExp(regexp).test(value)) {
        throw new Error(
          `passed parameter for ${paramName} should satisfy regxep: ${regexp}, got: ${value}`
        );
      }

      return prefix + value + postfix || part;
    }

    return part;
  }, parts).join('/');

  return normalizeTrailingSlash(`/${result}`, pathname.endsWith('/'));
};
