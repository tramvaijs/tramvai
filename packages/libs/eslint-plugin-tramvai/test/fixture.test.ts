import { runLintWithFixtures } from './utils/runLintWithFixtures';

const testCases = [
  {
    name: 'lint tramvai app files, configure tramvai plugin manually',
    type: 'tramvai',
    eslintConfig: {
      parser: 'babel-eslint',
      plugins: ['@tinkoff/tramvai'],
      rules: {
        '@tinkoff/tramvai/bundle-chunk-name': 'error',
      },
    },
  },
  {
    name: 'lint tramvai ts app files, use recommended tramvai plugin config',
    type: 'tramvai-typescript',
    eslintConfig: {
      parser: '@typescript-eslint/parser',
      extends: ['plugin:@tinkoff/tramvai/recommended'],
    },
  },
];

describe('fixture eslint plugin test', () => {
  testCases.forEach((testCase) => {
    it(`check: ${testCase.name}`, () => {
      const result = runLintWithFixtures(testCase.type, testCase.eslintConfig);
      expect(result).toMatchSnapshot();
    });
  });
});
