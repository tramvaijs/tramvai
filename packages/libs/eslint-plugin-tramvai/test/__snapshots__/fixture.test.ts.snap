// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixture eslint plugin test check: lint tramvai app files, configure tramvai plugin manually 1`] = `
{
  "error.js": {
    "errors": [
      "@tinkoff/tramvai/bundle-chunk-name",
      "@tinkoff/tramvai/bundle-chunk-name",
    ],
  },
  "ok.js": {},
}
`;

exports[`fixture eslint plugin test check: lint tramvai ts app files, use recommended tramvai plugin config 1`] = `
{
  "bundle1.ts": {},
  "bundle2.ts": {},
  "error.ts": {
    "errors": [
      "@tinkoff/tramvai/bundle-chunk-name",
      "@tinkoff/tramvai/bundle-chunk-name",
    ],
  },
  "ok.ts": {},
}
`;
